README.md
# Marcalyx

[![PyPi Version](https://badge.fury.io/py/marcalyx.svg)][pypi]
[![Build Status](http://img.shields.io/travis/seanredmond/marcalyx.svg)][travis]

[travis]: http://travis-ci.org/seanredmond/marcalyx
[pypi]: https://pypi.org/project/marcalyx


Simple python interface for reading
[MARC-XML](https://www.loc.gov/standards/marcxml/)


## Installation

    pip install marcalyx
    
## Usage

Import the package

    >>> import marcalyx

Marcalyx works on `xml.etree.ElementTree.Element` objects (or compatible), so
you will need to parse the XML before passing it to a Marcalyx object. For
instance, given a file containing a single `<record>`:

    >>> import xml.etree.ElementTree as ET
    
or, using `lxml`:

    >>> import lxml.etree as ET
    
then:

    >>> tree = ET.parse('tests/xml/1027474578.xml')
    >>> record_element = tree.getroot()
    >>> marc = marcalyx.Record(record_element)

If the XML contains a `<collection>` of Records you would instead create a `marcalyx.Collection` object, from which you can get the records:

    >>> coll = marcalyx.Collection(collection_element)
    >>> marc = coll.records()[0]

## Fields

Once you have a record, you can access the fields by tag (always use the 3-character string, for example, "001", "010", "100"):

    >>> marc.field("245")
    [245 10$aKindred /$cOctavia E. Butler.]
    
Or, more simply:

    >>> marc["245"]
    [245 10$aKindred /$cOctavia E. Butler.]
    
Fields are always returned as an array when accessed by tag (but not necessarily from the convenience methods, below). Fields are either a ControlField:

    >>> type(m["008"][0])
    <class 'marcalyx.marcalyx.ControlField'>
    
Or a DataField:

    >>> type(m["245"][0])
    <class 'marcalyx.marcalyx.DataField'>

All fields have a `tag` and a `value`:

    >>> m["008"][0].tag
    '008'
    >>> m["245"][0].tag
    '245'
    >>> m["008"][0].value
    '180306r20141979xxk    g      000 j eng d'
    >>> m["245"][0].value
    'Kindred / Octavia E. Butler.'

Data fields have their two "indicators":

    >>> m["245"][0].ind1
    '1'
    >>> m["245"][0].ind2
    '0'
    
And subfields, which can be accessed via the `subfield()` method, which returns an array:

    >>> m["245"][0].subfield('a')
    [$aKindred /]
    
As with fields, you can get subfields via subscript:

    >>> m["245"][0]['a']
    [$aKindred /]
    
You can use a tuple containing a field tag and subfield code as a subscript to
get a flat list of all the subfields with that code (if any) of all the fields
with that tag (if any):

    >>> m[('650','a')]
    [$aAfrican American women, $aTime travel]

## `value` vs. `str`

The string representation of a field is formatted in the conventional way, showing the indicators and subfields:

    >>> str(m["245"][0])
    '245 10$aKindred /$cOctavia E. Butler.'
    
The `value` is formatted for display:

    >>> m["245"][0].value
    'Kindred / Octavia E. Butler.'

## Subfields

Subfields have codes, values, and string representations:

    >>> m["245"][0].subfield("a")[0].code
    'a'
    >>> m["245"][0].subfield("a")[0].value
    'Kindred /'
    >>> str(m["245"][0].subfield("a")[0])
    '$aKindred /'
    
## Convenience methods

There are several methods to make it easier to get single fields or categories
of fields. `mainEntry()` will return whichever of the 1XX fields the record has
(as a `DataField`, not an array):

    >>> m.mainEntry()
    100 1#$aButler, Octavia Estelle$d(1947-2006).$4aut

`#titleStatement` gets the 245 field (again, as a `DataField` and not an array):

    >>> m.titleStatement()
    245 10$aKindred /$cOctavia E. Butler.

There are also methods to get an array of each of the main categories of
fields. Each of these returns an array of all the fields in the record of the
given category:

    >>> marc.controlFields() # 00X
    >>> marc.codes()         # 01X-09X
    >>> marc.titles()        # 20X-24X
    >>> marc.edition()       # 25X-28X
    >>> marc.description()   # 3XX
    >>> marc.series()        # 4XX
    >>> marc.notes()         # 5XX
    >>> marc.subjects()      # 6XX
    >>> marc.addedEntries()  # 70X-75X
    >>> marc.linking()       # 76X-78X
    >>> marc.seriesAdded()   # 80X-83X
    >>> marc.holdings()      # 841-88X

Some common numbers have convenience methods:

    > record.lccn       # 010$a, String or nil
    > record.isbns      # 020$a, Array of Strings, or []
    > record.issns      # 022$a, Array of Strings, or []

## Leader

You can get the record leader:

    >>> marc.leader
    '00000cam a2200000Mi 4500'

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/seanredmond/marcalyx. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [Contributor Covenant](http://contributor-covenant.org) code of conduct.

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).

## Code of Conduct

Everyone interacting in the Lccnorm project’s codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/seanredmond/marcalyx/blob/master/CODE_OF_CONDUCT.md).


CHANGELOG.md
# Changelog

## [1.0.2] - 2019-03-20
### Added
- Subscripting of data fields to get subfields
- Using tuple (field, subfield) as record subscript to get flat list of
  matching subfields of matching fields

## [1.0.1] - 2019-03-16
### Added
- Empty tags now pruned [@mwbenowitz](https://github.com/mwbenowitz)


tests/tests.py
import marcalyx
import pytest
import xml.etree.ElementTree as ET

@pytest.fixture()
def kindred():
    tree = ET.parse('tests/xml/1027474578.xml')
    root = tree.getroot()
    return marcalyx.Record(root)


@pytest.fixture()
def quilt():
    tree = ET.parse('tests/xml/10705.xml')
    root = tree.getroot()
    return marcalyx.Record(root)


@pytest.fixture()
def binti():
    tree = ET.parse('tests/xml/973807354.xml')
    root = tree.getroot()
    return marcalyx.Record(root)


@pytest.fixture()
def wrinkle():
    tree = ET.parse('tests/xml/14026028.xml')
    root = tree.getroot()
    return marcalyx.Record(root[0])


@pytest.fixture()
def marner():
    tree = ET.parse('tests/xml/2971.xml')
    root = tree.getroot()
    return marcalyx.Record(root)

@pytest.fixture()
def russian():
    tree = ET.parse('tests/xml/528635.xml')
    root = tree.getroot()
    return marcalyx.Record(root)


@pytest.fixture()
def xenophon():
    tree = ET.parse('tests/xml/3863.xml')
    root = tree.getroot()
    return marcalyx.Record(root)


@pytest.fixture()
def fissures():
    tree = ET.parse('tests/xml/53998.xml')
    root = tree.getroot()
    return marcalyx.Record(root)


def test_leader(kindred):
    assert kindred.leader == '00000cam a2200000Mi 4500'

    
def test_fields(kindred):
    f = kindred.fields
    ftypes = list(set([type(field) for field in kindred.fields]))
    assert isinstance(f, list)
    assert len(ftypes) == 2
    assert marcalyx.marcalyx.ControlField in ftypes
    assert marcalyx.marcalyx.DataField in ftypes


def test_subfield(kindred):
    s = kindred.subfield('650', 'a')
    assert isinstance(s, list)
    assert len(s) == 2
    assert list(set([type(sub) for sub in s])) == [marcalyx.marcalyx.SubField]


def test_subfield_when_some_are_empty(kindred):
    # There is one 651 field with an $x and one without
    s = kindred.subfield("651", "x")
    assert isinstance(s, list)
    assert len(s) == 1


def test_subfields_when_all_should_be_empy(kindred):
    s = kindred.subfield("650", "9")
    assert isinstance(s, list)
    assert len(s) == 0


def test_getting_a_field(kindred):
    f = kindred.field('245')
    assert isinstance(f, list)
    assert len(f) == 1
    assert f[0].tag == '245'


def test_getting_field_by_index(kindred):
    assert kindred.field('245') == kindred['245']

    
def test_getting_control_and_data_fields(kindred):
    assert isinstance(kindred['001'][0], marcalyx.marcalyx.ControlField)
    assert isinstance(kindred['245'][0], marcalyx.marcalyx.DataField)


def test_title_statement(kindred):
    assert str(kindred.titleStatement()) == \
        '245 10$aKindred /$cOctavia E. Butler.'


def test_datafield_value(kindred):
    assert kindred.titleStatement().value() == 'Kindred / Octavia E. Butler.'


def test_main_entry(kindred, quilt):
    k = kindred.mainEntry()
    q = quilt.mainEntry()

    assert isinstance(k, marcalyx.marcalyx.DataField)
    assert k.tag == '100'

    assert isinstance(q, marcalyx.marcalyx.DataField)
    assert q.tag == '111'


def test_control_fields(kindred):
    ctrl = kindred.controlFields()
    assert isinstance(ctrl, list)
    assert list(set([type(f) for f in ctrl])) == \
        [marcalyx.marcalyx.ControlField]
    assert len(ctrl) == 2
    assert ctrl[0],tag == '001'
    assert ctrl[-1].tag == '008'


def test_codes(kindred):
    codes = kindred.codes()
    assert isinstance(codes, list)
    assert len(codes) == 2
    assert codes[0].tag == "020"
    assert codes[0].value() == '9781472214812'


def test_titles(binti):
    t = binti.titles()
    assert isinstance(t, list)
    assert len(t) == 2
    assert t[0].tag == '245'
    assert t[-1].tag == '246'
    assert t[0].value() == 'Binti : the night masquerade / Nnedi Okorafor.'
    assert t[-1].value() == 'Night masquerade'


def test_editions(wrinkle):
    edition = wrinkle.edition()
    assert isinstance(edition, list)
    assert len(edition) ==  1
    assert edition[0].tag == '260'
    assert edition[0].value() == \
        'Harmondsworth, Eng. : Penguin Books, 1967, c1962.'


def test_description(wrinkle):
    description = wrinkle.description()
    assert isinstance(description, list)
    assert len(description) == 1
    assert description[0].tag == '300'
    assert description[0].value() == '184 p. ; 18 cm.'


def test_series(marner):
    series = marner.series()
    assert isinstance(series, list)
    assert len(series) == 1
    assert series[0].tag == '490'
    assert series[0].value() == 'Standard literature series ; no. 43'


def test_notes(quilt):
    notes = quilt.notes()
    assert isinstance(notes, list)
    assert len(notes) == 3
    assert notes[0].tag == '500'
    assert notes[0].value() == '"Millennium celebration."'


def test_subjects(kindred):
    subjects = kindred.subjects()
    assert isinstance(subjects, list)
    assert len(subjects) == 4
    assert subjects[0].tag == '650'
    assert subjects[0].value() == 'African American women Fiction.'


def test_added_entries(quilt):
    added = quilt.addedEntries()
    assert isinstance(added, list)
    assert len(added) == 2
    assert added[0].tag == '700'
    assert added[-1].tag == '710'
    assert added[0].value() == 'Browning, Bonnie K., 1944-'
    assert added[-1].value() == 'American Quilter\'s Society.'


def test_linking(russian):
    links = russian.linking()
    assert isinstance(links, list)
    assert len(links) == 1
    assert links[0].tag == '776'
    assert links[0].value() == 'Original (DLC)   61060095'


def test_series_added(xenophon):
    series = xenophon.seriesAdded()
    assert len(series) == 1
    assert series[0].tag == '810'
    assert series[0].value() == \
        'Cornell University. Cornell studies in classical philology ; no. 11.'


def test_holdings(binti):
    holdings = binti.holdings()
    assert len(holdings) == 1
    assert holdings[0].tag == '880'
    assert holdings[0].value() == '264-00 © 2017'


def test_lccn(quilt, kindred):
    assert quilt.lccn() == "   00010705 "
    assert kindred.lccn() is None
    

def test_isbns(kindred):
    assert isinstance(kindred.isbns(), list)
    assert kindred.isbns() == ['9781472214812', '1472214811']


def test_issns(fissures):
    assert isinstance(fissures.issns(), list)
    assert fissures.issns() == ['1572733691 (pbk.)']


tests/test_record.py
# -*- coding: utf-8 -*-

import marcalyx
import pytest
import xml.etree.ElementTree as ET

@pytest.fixture()
def kindred():
    tree = ET.parse('tests/xml/1027474578.xml')
    root = tree.getroot()
    return marcalyx.Record(root)


@pytest.fixture()
def quilt():
    tree = ET.parse('tests/xml/10705.xml')
    root = tree.getroot()
    return marcalyx.Record(root)


@pytest.fixture()
def binti():
    tree = ET.parse('tests/xml/973807354.xml')
    root = tree.getroot()
    return marcalyx.Record(root)


@pytest.fixture()
def wrinkle():
    tree = ET.parse('tests/xml/14026028.xml')
    root = tree.getroot()
    return marcalyx.Record(root[0])


@pytest.fixture()
def marner():
    tree = ET.parse('tests/xml/2971.xml')
    root = tree.getroot()
    return marcalyx.Record(root)

@pytest.fixture()
def russian():
    tree = ET.parse('tests/xml/528635.xml')
    root = tree.getroot()
    return marcalyx.Record(root)


@pytest.fixture()
def xenophon():
    tree = ET.parse('tests/xml/3863.xml')
    root = tree.getroot()
    return marcalyx.Record(root)


@pytest.fixture()
def fissures():
    tree = ET.parse('tests/xml/53998.xml')
    root = tree.getroot()
    return marcalyx.Record(root)


@pytest.fixture()
def tokio():
    tree = ET.parse('tests/xml/26003.xml')
    root = tree.getroot()
    return marcalyx.Record(root)



def test_leader(kindred):
    assert kindred.leader == '00000cam a2200000Mi 4500'

    
def test_subfield(kindred):
    s = kindred.subfield('650', 'a')
    assert isinstance(s, list)
    assert len(s) == 2
    assert list(
        set([isinstance(sub, marcalyx.marcalyx.SubField) for sub in s])
    ) == [True]


def test_subfield_when_some_are_empty(kindred):
    # There is one 651 field with an $x and one without
    s = kindred.subfield("651", "x")
    assert isinstance(s, list)
    assert len(s) == 1


def test_subfields_when_all_should_be_empy(kindred):
    s = kindred.subfield("650", "9")
    assert isinstance(s, list)
    assert len(s) == 0


def test_self_closing_subfield(tokio):
    es = tokio.subfield("520", "a")
    assert isinstance(es, list)
    assert len(es) == 0


def test_getting_a_field(kindred):
    f = kindred.field('245')
    assert isinstance(f, list)
    assert len(f) == 1
    assert f[0].tag == '245'


def test_getting_field_by_index(kindred):
    assert kindred.field('245') == kindred['245']


def test_getting_subfield_by_index(kindred):
    subf = kindred[('650', 'a')]
    assert isinstance(subf, list)
    assert len(subf) == 2
    assert subf[0].value == 'African American women'

    
def test_getting_control_and_data_fields(kindred):
    assert isinstance(kindred['001'][0], marcalyx.marcalyx.ControlField)
    assert isinstance(kindred['245'][0], marcalyx.marcalyx.DataField)


def test_title_statement(kindred):
    assert str(kindred.titleStatement()) == \
        '245 10$aKindred /$cOctavia E. Butler.'


def test_datafield_value(kindred):
#    assert kindred.titleStatement().value == 'Kindred / Octavia E. Butler.'
    assert kindred.titleStatement().value == 'Kindred / Octavia E. Butler.'


def test_main_entry(kindred, quilt):
    k = kindred.mainEntry()
    q = quilt.mainEntry()

    assert isinstance(k, marcalyx.marcalyx.DataField)
    assert k.tag == '100'

    assert isinstance(q, marcalyx.marcalyx.DataField)
    assert q.tag == '111'


def test_control_fields(kindred):
    ctrl = kindred.controlFields()
    assert isinstance(ctrl, list)
    assert list(
        set([isinstance(f, marcalyx.marcalyx.ControlField) for f in ctrl])
    ) == [True]
    assert len(ctrl) == 2
    assert ctrl[0],tag == '001'
    assert ctrl[-1].tag == '008'


def test_codes(kindred):
    codes = kindred.codes()
    assert isinstance(codes, list)
    assert len(codes) == 2
    assert codes[0].tag == "020"
    assert codes[0].value == '9781472214812'


def test_titles(binti):
    t = binti.titles()
    assert isinstance(t, list)
    assert len(t) == 2
    assert t[0].tag == '245'
    assert t[-1].tag == '246'
    assert t[0].value == 'Binti : the night masquerade / Nnedi Okorafor.'
    assert t[-1].value == 'Night masquerade'


def test_editions(wrinkle):
    edition = wrinkle.edition()
    assert isinstance(edition, list)
    assert len(edition) ==  1
    assert edition[0].tag == '260'
    assert edition[0].value == \
        'Harmondsworth, Eng. : Penguin Books, 1967, c1962.'


def test_description(wrinkle):
    description = wrinkle.description()
    assert isinstance(description, list)
    assert len(description) == 1
    assert description[0].tag == '300'
    assert description[0].value == '184 p. ; 18 cm.'


def test_series(marner):
    series = marner.series()
    assert isinstance(series, list)
    assert len(series) == 1
    assert series[0].tag == '490'
    assert series[0].value == 'Standard literature series ; no. 43'


def test_notes(quilt):
    notes = quilt.notes()
    assert isinstance(notes, list)
    assert len(notes) == 3
    assert notes[0].tag == '500'
    assert notes[0].value == '"Millennium celebration."'


def test_subjects(kindred):
    subjects = kindred.subjects()
    assert isinstance(subjects, list)
    assert len(subjects) == 4
    assert subjects[0].tag == '650'
    assert subjects[0].value == 'African American women Fiction.'


def test_added_entries(quilt):
    added = quilt.addedEntries()
    assert isinstance(added, list)
    assert len(added) == 2
    assert added[0].tag == '700'
    assert added[-1].tag == '710'
    assert added[0].value == 'Browning, Bonnie K., 1944-'
    assert added[-1].value == 'American Quilter\'s Society.'


def test_linking(russian):
    links = russian.linking()
    assert isinstance(links, list)
    assert len(links) == 1
    assert links[0].tag == '776'
    assert links[0].value == 'Original (DLC)   61060095'


def test_series_added(xenophon):
    series = xenophon.seriesAdded()
    assert len(series) == 1
    assert series[0].tag == '810'
    assert series[0].value == \
        'Cornell University. Cornell studies in classical philology ; no. 11.'


def test_holdings(binti):
    holdings = binti.holdings()
    assert len(holdings) == 1
    assert holdings[0].tag == '880'
    assert holdings[0].value == u'264-00 © 2017'



def test_lccn(quilt, kindred):
    assert quilt.lccn() == "   00010705 "
    assert kindred.lccn() is None
    

def test_isbns(kindred):
    assert isinstance(kindred.isbns(), list)
    assert kindred.isbns() == ['9781472214812', '1472214811']


def test_issns(fissures):
    assert isinstance(fissures.issns(), list)
    assert fissures.issns() == ['1572733691 (pbk.)']


tests/test_datafield.py
import marcalyx
import pytest
import xml.etree.ElementTree as ET


@pytest.fixture()
def dataf():
    tree = ET.parse('tests/xml/1027474578.xml')
    root = tree.getroot()
    return marcalyx.Record(root)['100'][0]


def test_tag(dataf):
    assert dataf.tag == '100'


def test_value(dataf):
    assert dataf.value == 'Butler, Octavia Estelle (1947-2006). aut'


def test_to_string(dataf):
    assert str(dataf) == '100 1#$aButler, Octavia Estelle$d(1947-2006).$4aut'


def test_subfields(dataf):
    assert isinstance(dataf.subfields, list)
    assert len(dataf.subfields) == 3


def test_subfield(dataf):
    subf = dataf.subfield('a')
    assert isinstance(subf, list)
    assert isinstance(subf[0], marcalyx.marcalyx.SubField)
    assert subf[0].value == 'Butler, Octavia Estelle'


def test_subfield_getitem(dataf):
    subf = dataf['a']
    assert isinstance(subf, list)
    assert isinstance(subf[0], marcalyx.marcalyx.SubField)
    assert subf[0].value == 'Butler, Octavia Estelle'


marcalyx/marcalyx.py
import re


class MarcNamespacedElement:
    SLIM = {'marc': 'http://www.loc.gov/MARC21/slim'}

    def get_namespace(self, node):
        ns_search = re.search('({(.+)})?.+', node.tag)
        if ns_search.group(2) == 'http://www.loc.gov/MARC21/slim':
            return ns_search.group(2)

        if ns_search.group(2) is None:
            return None

        raise UnrecognizedNamespaceError(ns_search.group(2))

    def find_with_ns(self, node, tag, ns):
        if ns:
            ns_str = 'marc:'
        else:
            ns_str = ''
        return node.findall("%s%s" % (ns_str, tag,), self.SLIM)


class Collection(MarcNamespacedElement):
    def __init__(self, node):
        self.node = node

    def records(self):
        return [Record(r) for r in
                self.find_with_ns(self.node, 'record',
                                  self.get_namespace(self.node))]
    
class Record(MarcNamespacedElement):
    def __init__(self, node):
        self.node = node

        ns = self.get_namespace(node)

        self.leader = self.get_leader(node, ns)
        self.fields = list(filter(None,
            [ControlField(f) for f in self.find_control(node, ns)] + \
            [DataField(f, ns) for f in self.find_data(node, ns)]
        ))
        
    def find_control(self, node, ns):
        return self.find_with_ns(node, 'controlfield', ns)

    def find_data(self, node, ns):
        return self.find_with_ns(node, 'datafield', ns)

    def get_leader(self, node, ns):
        leader = self.find_with_ns(node, 'leader', ns)
        return leader[0].text

    def mainEntry(self):
        return next(iter(self.tag_range("100", "1XX") or []), None)

    def titleStatement(self):
        return self.field('245')[0]

    def controlFields(self):
        return self.tag_range('001', '009')

    def codes(self):
        """Return the numbers and code fields (01X-09X)."""

        return self.tag_range('010', '09X')

    def titles(self):
        """Return the title- and title-related fields (20X-24X)."""

        return self.tag_range('200', '24X')

    def edition(self):
        """Return the edition- and imprint-related fields (25X-28X)."""

        return self.tag_range('250', '28X')

    def description(self):
        """Return the physical description fields (3XX)."""

        return self.tag_range('300', '399')

    def series(self):
        """Return the series statement fields (4XX)."""

        return self.tag_range('400', '499')

    def notes(self):
        """Return the note statement fields (5XX)."""

        return self.tag_range('500', '59X')

    def subjects(self):
        """Return the subject access fields (6XX)."""

        return self.tag_range('600', '69X')

    def addedEntries(self):
        """Return the added entry fields (70X-75X)."""

        return self.tag_range('700', '75X')

    def linking(self):
        """Return the linking entry fields (76X-78X)."""

        return self.tag_range('760', '78X')

    def seriesAdded(self):
        """Return the series added entry fields (80X-83X)."""

        return self.tag_range('800', '83X')

    def holdings(self):
        """Return the holdings, alternate graphics, etc fields (841-88X)"""

        return self.tag_range('841', '88X')

    def lccn(self):
        """Return the Library of Congress Control Number.

        Returns the Library of Congress Control Number (010$a). Only returns
        the $a subfields, nothing from the $b or $z. Returns None if there is
        no LCCN in the record."""

        s = self.subfield('010', 'a')
        if len(s):
            return s[0].value

        return None

    def isbns(self):
        """Return an array of International Standard Book Numbers (020$a)."""

        return [s.value for s in self.subfield('020', 'a')]

    def issns(self):
        """Return an array of International Standard Serial Numbers (022$a)."""

        return [s.value for s in self.subfield('022', 'a')]

    def field(self, tag):
        return [f for f in self.fields if f.tag == tag]

    def subfield(self, tag, code):
        return [g for h in
                [f.subfield(code) for f in self.field(tag)]
                for g in h]

    def tag_range(self, first, last):
        return [f for f in self.fields if f.tag >= first and f.tag <= last]

    def __getitem__(self, tag):
        if isinstance(tag, tuple):
            return [a for b in [f.subfield(tag[1]) \
                                for f in self.field(tag[0])] for a in b]

        return self.field(tag)


class ControlField:
    def __init__(self, node):
        self.node = node
        self.tag = node.attrib['tag']
        self.value = node.text
        self.subfields = []

    def __repr__(self):
        return "%s   %s" % (self.tag, self.value,)


class DataField(MarcNamespacedElement):
    def __init__(self, node, ns):
        self.node = node
        self.tag = node.attrib['tag']
        self.ind1 = node.attrib['ind1']
        self.ind2 = node.attrib['ind2']
        self.subfields = [
            SubField(s) for s in
            self.find_with_ns(node, 'subfield', ns)
            if s.text is not None
        ]
        self.value = ' '.join([s.value for s in self.subfields]) \
                      if len(self.subfields) > 0 else None

    def subfield(self, code):
        return [s for s in self.subfields if s.code == code]

    def ind_to_s(self, i):
        if i == ' ':
            return '#'

        return i

    def __getitem__(self, code):
        return self.subfield(code)


    def __repr__(self):
        return "%s %s%s%s" % (self.tag,
                              self.ind_to_s(self.ind1),
                              self.ind_to_s(self.ind2),
                              ''.join([str(s) for s in self.subfields]),)


class SubField:
    def __init__(self, node):
        self.code = node.attrib['code']
        self.value = node.text

    def __repr__(self):
        return "$%s%s" % (self.code, self.value,)


class UnrecognizedNamespaceError(Exception):
    def __init__(self, value):
        self.value = value

    def __str__(self):
        return repr(self.value)
